xwiki-platform-core/xwiki-platform-resource/xwiki-platform-resource-servlet/src/main/java/org/xwiki/resource/servlet/AbstractServletResourceReferenceHandler.java

```diff
@@ -174,15 +175,33 @@ private void serveResource(R resourceReference, InputStream rawResourceStream)
         try {
             Response response = this.container.getResponse();
             setResponseHeaders(response, resourceReference);
-            response.setContentType(TikaUtils.detect(resourceStream, resourceName));
+            response.setContentType(getContentType(resourceStream, resourceReference));
             IOUtils.copy(resourceStream, response.getOutputStream());
         } catch (Exception e) {
+            String resourceName = getResourceName(resourceReference);
             throw new ResourceReferenceHandlerException(String.format("Failed to read resource [%s]", resourceName), e);
         } finally {
             IOUtils.closeQuietly(resourceStream);
         }
     }
 
+    /**
+     * Computes the content type of the resource. By default the content type is inferred by {@link
+     * TikaUtils#detect(InputStream, String)} based on the resource content and name.
```

Comment:
Maybe the javadoc should mention that this is very costly and if the content type is known, then the RRH should set it,WDYT?

Reply:
> this can be very costly

Why?

Reply:
Because reading the stream content to infer the type is costly.